receiver/lokireceiver/internal/util.go

```diff
@@ -0,0 +1,136 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/lokireceiver/internal"
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+)
+
+// CompressionType for encoding and decoding requests and responses.
+type CompressionType int
+
+// Values for CompressionType
+const (
+	NoCompression CompressionType = iota
+	RawSnappy
+)
+
+const messageSizeLargerErrFmt = "received message larger than max (%d vs %d)"
+
+// parseProtoReader parses a compressed proto from an io.Reader.
+func parseProtoReader(reader io.Reader, expectedSize, maxSize int, req proto.Message, compression CompressionType) error {
+	body, err := decompressRequest(reader, expectedSize, maxSize, compression)
+	if err != nil {
+		return err
+	}
+
+	// We re-implement proto.Unmarshal here as it calls XXX_Unmarshal first,
+	// which we can't override without upsetting golint.
+	req.Reset()
+	if u, ok := req.(proto.Unmarshaler); ok {
+		err = u.Unmarshal(body)
+	} else {
+		err = proto.NewBuffer(body).Unmarshal(req)
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func decompressRequest(reader io.Reader, expectedSize, maxSize int, compression CompressionType) (body []byte, err error) {
+	defer func() {
+		if err != nil && len(body) > maxSize {
+			err = fmt.Errorf(messageSizeLargerErrFmt, len(body), maxSize)
+		}
+	}()
+	if expectedSize > maxSize {
+		return nil, fmt.Errorf(messageSizeLargerErrFmt, expectedSize, maxSize)
+	}
+	buffer, ok := tryBufferFromReader(reader)
+	if ok {
+		body, err = decompressFromBuffer(buffer, maxSize, compression)
+		return
+	}
+	body, err = decompressFromReader(reader, expectedSize, maxSize, compression)
+	return
```

Comment:
It is "naked" return. It is needed because the function returns named values `body`, `err`.
`return` in the named function is a shortcut for `return body, err`